writer.js

```diff
@@ -0,0 +1,78 @@
+const nsqStream = require('nsq-stream');
+
+/**
+ * @constructor
+ * @param {Object} opts - options for the writer
+ */
+function Writer(opts = {}) {
+  const { writer, topic, spec = {}, log } = opts;
+  this.log = log;
+  this.writeStream = writer && topic && nsqStream.createWriteStream(writer, topic);
+  this.spec = spec;
+}
+
+/**
+ * Combines the spec with status information
+ *
+ * @function buildStatusMessage
+ * @param {Object} statusInfo - Status information
+ * @returns {Object} - The combined message object
+ * @api private
+ */
+Writer.prototype.buildStatusMessage = function buildStatusMessage(statusInfo) {
+  const { type: buildType, ...other } = this.spec;
+  return {
+    ...statusInfo,
+    ...other,
+    buildType
+  };
+};
+
+/**
+ * @function _doStreamAction
+ * @param {String} action - Which stream action execute
+ * @param {Object} statusInfo - Status information
+ * @param {Function} done - Callback
+ * @returns {undefined}
+ * @api private
+ */
+Writer.prototype._doStreamAction = function _doStreamAction(action, statusInfo, done) {
+  if (!this.writeStream) return done();
+
+  const msg = this.buildStatusMessage(statusInfo);
+
+  if (this.writeStream._writableState.ended) {
+    this.log.error(`Unable to ${action === 'write' ? 'write to' : action} stream`, msg);
+    return done();
+  }
+
+  this.writeStream[action](msg, done);
+};
+
+/**
+ * @function write
+ * @param {Error} err - Error object for unsuccessful actions
+ * @param {Object} statusInfo - Status information
+ * @param {Function} done - Callback
+ * @api public
+ */
+Writer.prototype.write = function write(err, statusInfo = {}, done = () => {}) {
+  if (err) {
```

Comment:
idk if its just me but having `err` as first argument to a stream-like object seems off. Can we just specify the eventType on the statusInfo that we pass in on an error? Or set the message to the error object itself and unwrap it in here? We will probably want to handle making sure we send the properties attached to any error object. This could all be handled in `buildStatusMessage` so the weirdness isn't put on the caller of this function

Reply:
I had it that way (setting it in the status opbject) originally, and it produced a bunch of duplicate code. I can switch back to it though.

Reply:
What about option 3? Putting the `error` in the `statusInfo`.